pkg/controllers/machine_controller.go

```diff
@@ -268,35 +265,29 @@ func (r *MachineReconciler) startCheckAndEnqueueVolumeResize(ctx context.Context
 }
 
 func (r *MachineReconciler) startCheckAndEnqueueMachineState(ctx context.Context, log logr.Logger) {
-	wait.UntilWithContext(ctx, func(ctx context.Context) {
-		machines, err := r.machines.List(ctx)
-		if err != nil {
-			log.Error(err, "failed to list machines")
-			return
-		}
+	eventChan, err := r.libvirt.LifecycleEvents(ctx)
+	if err != nil {
+		log.Error(err, "failed to subscribe lifecycle events")
+		return
+	}
 
-		for _, machine := range machines {
-			if machine.DeletedAt != nil || !slices.Contains(machine.Finalizers, MachineFinalizer) {
-				continue
-			}
-			id := machine.ID
+	log.Info("subscribing lifecycle events")
 
-			state, err := r.getMachineState(id)
-			if err != nil {
-				if libvirt.IsNotFound(err) {
-					log.V(1).Info("Failed to retrieve domain. Requeueing", "machineID", id)
-					r.queue.AddRateLimited(id)
-				} else {
-					log.Error(err, "failed to get machine state", "machineID", id)
-				}
-				continue
+	for {
```

Comment:
This can be simplified. 
```suggestion
	for ev := range eventChan {
		log.Info("requeue machine id " + ev.Dom.Name)
		r.queue.AddRateLimited(ev.Dom.Name)
	}

	if err := ctx.Err(); err != nil && err != context.Canceled {
		log.Error(err, "context error")
	}
```

Reply:
you have to first process all events in channel after you can exit. I don't see big benefits in trigger requeue when main context is close and it cannot be processes anymore.

Reply:
ok. makes sense.